manager/allocator/network/port/port.go

```diff
@@ -0,0 +1,470 @@
+		if spec.PublishedPort != 0 {
+			continue
+		}
+		// check if there is an old spec. If not, we don't have to go looking
+		// for a previously assigned dynamic port
+		if endpoint.Spec == nil {
+			continue
+		}
+		// go through all of the ports in the old spec, checking if one matches
+		for _, old := range endpoint.Spec.Ports {
+			if portsEqual(old, spec) {
+				// if the do match, find the port from the object
+				for _, p := range endpoint.Ports {
+					if PortsMostlyEqual(spec, p) {
+						// and take it's PublishedPort assignment for the final
+						// port assignment
+						finalPorts[i].PublishedPort = p.PublishedPort
+					}
+				}
+			}
+		}
+	}
+
+	// to avoid altering the "real" state of the ports map, we use the proposal
+	// object, which contains the proposed changes to the Allocator. this
+	// means if there is a failure in the caller after calling Allocate, the
+	// caller can discard the changes
+	prop := &proposal{
+		pa:         pa,
+		allocate:   make(map[port]struct{}, len(finalPorts)),
+		deallocate: make(map[port]struct{}, len(endpoint.Ports)),
+	}
+
+	// now, deallocate everything in the old object
+	for _, p := range endpoint.Ports {
+		// skip publish mode ports
+		if p.PublishMode == api.PublishModeHost {
+			continue
+		}
+
+		prop.deallocate[port{p.Protocol, p.PublishedPort}] = struct{}{}
+	}
+
+	// and then allocate everything in the new object
+
+	// we'll do this in two steps. in the first step, allocate any ports that
+	// have a published port assigned already. this first step prevents us from
+	// choosing a published port for one port that another, later port wants to
+	// use
+	for _, p := range finalPorts {
+		// Skip all Host ports, which we take no action on
+		if p.PublishedPort == 0 || p.PublishMode == api.PublishModeHost {
+			continue
+		}
+		portObj := port{p.Protocol, p.PublishedPort}
+		if _, ok := pa.ports[portObj]; ok {
+			// check if we're deallocating this port
+			if _, ok := prop.deallocate[portObj]; !ok {
+				return nil, errors.ErrResourceInUse("port", portObj.String())
+			}
+		}
+		// also check that we haven't already tried to allocate this port for
+		// this particular endpoint
+		if _, ok := prop.allocate[portObj]; ok {
+			return nil, errors.ErrInvalidSpec("published port %v is assigned to more than 1 port config", portObj)
+		}
+
+		// now, mark this port as "in use" in the newPorts map.
+		prop.allocate[portObj] = struct{}{}
+	}
+
+	// now, this second go around, we'll choose all new publish ports to
+	// dynamically allocate for any port that doesn't have a publish port yet
+ports:
+	for _, p := range finalPorts {
+		// Again, skip all host ports
+		if p.PublishedPort != 0 || p.PublishMode == api.PublishModeHost {
+			continue
+		}
+		portObj := port{p.Protocol, DynamicPortStart}
+		// loop through the whole range of dynamic ports and select the first
+		// one available
+		for i := DynamicPortStart; i <= DynamicPortEnd; i++ {
```

Comment:
it's worst case O(n) with the number of ports, because everything is done off map lookups. There are more efficient ways to dynamically assign ports (@stevvooe has mentioned things like bit-twiddling) but I don't think they'll be worth optimizing unless we see that this behavior is bottlenecking performance.

Reply:
Made a PR to your `rewrite_allocator` branch with this optimization and another one focused on the loops that re-allocates deallocated ports. Was quite straight-forward.